nixos/modules/services/web-apps/stirling-pdf.nix

```diff
@@ -0,0 +1,110 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.stirling-pdf;
+in
+{
+  options.services.stirling-pdf = {
+    enable = lib.mkEnableOption "the stirling-pdf service";
+
+    package = lib.mkPackageOption pkgs "stirling-pdf" { };
+
+    environment = lib.mkOption {
+      type = lib.types.attrsOf (
+        lib.types.oneOf [
+          lib.types.str
+          lib.types.int
+        ]
+      );
+      default = { };
+      example = {
+        SERVER_PORT = 8080;
+        INSTALL_BOOK_AND_ADVANCED_HTML_OPS = "true";
+      };
+      description = ''
+        Environment variables for the stirling-pdf app.
+        See https://github.com/Stirling-Tools/Stirling-PDF#customisation for available options.
+      '';
+    };
+
+    environmentFiles = lib.mkOption {
+      type = lib.types.listOf lib.types.path;
+      default = [ ];
+      description = ''
+        Files containing additional environment variables to pass to Stirling PDF.
+        Secrets should be added in environmentFiles instead of environment.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.stirling-pdf = {
+      environment = lib.mapAttrs (_: toString) cfg.environment;
+
+      # following https://github.com/Stirling-Tools/Stirling-PDF#locally
+      path = with pkgs; [
+        unpaper
+        libreoffice
+        ocrmypdf
+        poppler_utils
+        unoconv
+        opencv
+        pngquant
+        tesseract
+        python3Packages.weasyprint
+        calibre
+      ];
```

Comment:
Why don't we patch the source instead by e.g. hardcoding the path to ocrmypdf in https://github.com/Stirling-Tools/Stirling-PDF/blob/d00e7fe958ac4c8fe315fa4891b7a46d16dbb8c9/src/main/java/stirling/software/SPDF/controller/api/converters/ConvertPDFToPDFA.java#L92?

Reply:
That's a good point. I'll work on a PR later for the stirling-pdf pkg itself

Reply:
I think making it part of the service is nice, as it could allow introducing a mechanism similar to what is done in [`services.paperless`](https://github.com/NixOS/nixpkgs/blob/c374d94f1536013ca8e92341b540eba4c22f9c62/nixos/modules/services/misc/paperless.nix#L204) to reduce the closure size of the service.

Similarly, `calibre` could depend on `INSTALL_BOOK_AND_ADVANCED_HTML_OPS`'s value.

Reply:
Just a note about patching the source of java programs:
`.jar` files are compressed, so references to `/nix/store/*` get forgotten, needing some workarounds (e.g. putting the dependencies' paths into comments in the main bash script).

Using `makeWrapper` with `--prefix PATH` can help usually with that.

We could, like mentioned above, wrap the package itself with these extra dependencies, instead of relying of the user/service to install them. Not sure which is better.